crates/re_log_encoding/src/decoder/stream.rs

```diff
@@ -0,0 +1,281 @@
+    use re_log_types::StoreId;
+    use re_log_types::StoreInfo;
+    use re_log_types::StoreKind;
+    use re_log_types::StoreSource;
+    use re_log_types::Time;
+
+    use crate::encoder::Encoder;
+    use crate::EncodingOptions;
+
+    use super::*;
+
+    fn fake_log_msg() -> LogMsg {
+        LogMsg::SetStoreInfo(SetStoreInfo {
+            row_id: RowId::ZERO,
+            info: StoreInfo {
+                application_id: ApplicationId::unknown(),
+                store_id: StoreId::from_string(StoreKind::Recording, "test".into()),
+                is_official_example: false,
+                started: Time::from_ns_since_epoch(0),
+                store_source: StoreSource::Unknown,
+                store_kind: StoreKind::Recording,
+            },
+        })
+    }
+
+    fn test_data(options: EncodingOptions, n: usize) -> Vec<u8> {
+        let mut buffer = Vec::new();
+        {
+            let mut encoder = Encoder::new(options, &mut buffer).unwrap();
+            for _ in 0..n {
+                encoder.append(&fake_log_msg()).unwrap();
+            }
+        }
+        buffer
+    }
+
+    macro_rules! assert_message_ok {
+        ($message:expr) => {{
+            match $message {
+                Ok(Some(message)) => {
+                    assert_eq!(&fake_log_msg(), &message)
+                }
+                Ok(None) => {
+                    panic!("failed to read message: message could not be read in full");
+                }
+                Err(err) => {
+                    panic!("failed to read message: {err}");
+                }
+            }
+        }};
+    }
+
+    #[test]
+    fn stream_whole_chunks_uncompressed() {
+        let data = test_data(EncodingOptions::UNCOMPRESSED, 16);
+
+        let mut decoder = StreamDecoder::new();
+        decoder.push_chunk(data);
+
+        for _ in 0..16 {
+            assert_message_ok!(decoder.try_read());
+        }
+    }
+
+    #[test]
+    fn stream_byte_chunks_uncompressed() {
+        let data = test_data(EncodingOptions::UNCOMPRESSED, 16);
+
+        let mut decoder = StreamDecoder::new();
+        for chunk in data.chunks(1) {
+            decoder.push_chunk(chunk.to_vec());
+        }
+
+        for _ in 0..16 {
+            assert_message_ok!(decoder.try_read());
+        }
+    }
+
+    #[test]
+    fn stream_whole_chunks_compressed() {
+        let data = test_data(EncodingOptions::COMPRESSED, 16);
+
+        let mut decoder = StreamDecoder::new();
+        decoder.push_chunk(data);
+
+        for _ in 0..16 {
+            assert_message_ok!(decoder.try_read());
+        }
+    }
+
+    #[test]
+    fn stream_byte_chunks_compressed() {
+        let data = test_data(EncodingOptions::COMPRESSED, 16);
+
+        let mut decoder = StreamDecoder::new();
+        for chunk in data.chunks(1) {
+            decoder.push_chunk(chunk.to_vec());
+        }
+
+        for _ in 0..16 {
+            assert_message_ok!(decoder.try_read());
+        }
+    }
```

Comment:
you should also test that you can:

* call `decoder.try_read()` without first pushing any chunks (and get `None`)
* push a few chunks, get a few things out of `decoder.try_read()`, then `None`, and then push the rest of the chunks and read the rest of the messages (i.e. actually stream!)
* push chunks of various sizes (not just the edge-case of size `1`, which never tests the `Cursor` part of `Chunk`